### DISCUSSION TITLE
Need help configuring the language server with neovim

### DISCUSSION THREAD

**Discussion by asher-gh** [znck/grammarly#285] [Q&A]
I wanted to configure the language server with nvim-lsp config. Where does the server look for config file or can this be passed as option to lspconfig?

Getting this error when using [nvim-lsp-installer](https://github.com/williamboman/nvim-lsp-installer/discussions/805)

```
Error executing vim.schedule lua callback: ...w/Cellar/neovim/0.7.0/share/nvim/runtime/lua/vim/
lsp.lua:955: RPC[Error] code_name = InternalError, message = "Request initialize failed with me
ssage: clientId is required"
stack traceback:
        [C]: in function 'assert'
        ...w/Cellar/neovim/0.7.0/share/nvim/runtime/lua/vim/lsp.lua:955: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```

**Comment by kofm** [comment]
I get the same error. Have you found any solution?

**Comment by tomrtk** [comment]
After digging some around the source code and docs I found if adding this line to the default `nvim-lspconfig`, solves this problem.

```lua
require('lspconfig').grammarly.setup{
	init_options = { clientId = 'client_BaDkMgx4X19X9UxxYRCXZo', },
}
```
with `clientId` from [here](https://github.com/znck/grammarly/blob/fb10db7e5a7bd23a49537aa899905e8222d079dd/extension/src/GrammarlyClient.ts#L71)

**Comment by weakish** [comment]
Thanks to @tomrtk , I added the following  configuration to LunarVim:

```lua
-- edit this file by running `:lua vim.cmd("edit " .. lvim.lsp.templates_dir .. "/markdown.lua")`
local grammarly_opts = {
  init_options = { clientId = 'client_BaDkMgx4X19X9UxxYRCXZo', },
}
require("lvim.lsp.manager").setup("grammarly", grammarly_opts)
```

And it works. :tada:

**Comment by kofm** [comment]
Still not working for me :( 
I get a cryptic "Client 1 quit with exit code 1 and signal 0"
How did you install the executable?

**Reply by tomrtk** [reply]
I am on nixos, but I installed it in an npm equivalent way. What happens if you start the server in a terminal? I get this output.

```command
❯ grammarly-languageserver --stdio
Content-Length: 85

{"jsonrpc":"2.0","method":"window/logMessage","params":{"type":4,"message":"Ready!"}}
```

When I was experimenting, I put this line in my `init.lua` config,

```lua
vim.lsp.set_log_level("debug")
```

and used the `:LspLog` command in nvim to see errors etc.

Maybe it will guide you in a direction of what is wrong.

**Reply by tomrtk** [reply]
Also I had a quick look at the `nvim-lsp-installer` grammarly definition, and I do not see how they pass the `--stdio` args to the executable. It was removed at some point. Also the project seems to be no longer maintained.

**Reply by weakish** [reply]
> Also the project seems to be no longer maintained.

Yes. You can consider migrating to mason.nvim

**Reply by DanRoscigno** [reply]
@kofm  this is how I set the --stdin, and the `Client 1 quit with exit code 1 and signal 0` stopped.  I don't know how to get Grammarly to check my document, but `:LspInfo` shows that it is ok.  
```
require('lspconfig').grammarly.setup{
       init_options = { clientId = 'client_BaDkMgx4X19X9UxxYRCXZo', },
       cmd = { "grammarly-languageserver", "--stdio" },
 }
```

Dows anyone know the commands to use Grammarly once it is running in nvim?

**Reply by kofm** [reply]
After tinkering without success, I stumbled upon a weird behaviour from Neovim. If the file includes a vi modeline as the last line, like the following:
```
<!-- vi:filetype=pandoc conceallevel=0
-->
```

It magically works... I don't understand why. (I have the same config reported by you and installed the lsp via mason.nvim)

EDIT:

The error is probably related to some clash with Neovim's tree-sitter. Within the logs, this is the error:

`[ERROR][2022-11-17 08:47:20] .../vim/lsp/rpc.lua:733	"rpc"	"grammarly-languageserver"	"stderr"	"node:internal/deps/undici/undici:14294\n    Error.captureStackTrace(err, this);\n          ^\n\nTypeError: Failed to parse URL from /home/kofm/.local/share/nvim/mason/packages/grammarly-languageserver/node_modules/grammarly-languageserver/node_modules/web-tree-sitter/tree-sitter.wasm\n    at Object.fetch (node:internal/deps/undici/undici:14294:11) {\n  [cause]: TypeError [ERR_INVALID_URL]: Invalid URL\n      at new NodeError (node:internal/errors:393:5)\n      at URL.onParseError (node:internal/url:565:9)\n      at new URL (node:internal/url:645:5)\n      at new Request (node:internal/deps/undici/undici:7023:25)\n      at fetch2 (node:internal/deps/undici/undici:13460:25)\n      at Object.fetch (node:internal/deps/undici/undici:14292:18)\n      at fetch (node:internal/process/pre_execution:238:25)\n      at /home/kofm/.local/share/nvim/mason/packages/grammarly-languageserver/node_modules/grammarly-languageserver/node_modules/web-tree-sitter/tree-sitter.js:1:15192\n      at /home/kofm/.local/share/nvim/mason/packages/grammarly-languageserver/node_modules/grammarly-languageserver/node_modules/web-tree-sitter/tree-sitter.js:1:15413\n      at new Promise (<anonymous>) {\n    input: '/home/kofm/.local/share/nvim/mason/packages/grammarly-languageserver/node_modules/grammarly-languageserver/node_modules/web-tree-sitter/tree-sitter.wasm',\n    code: 'ERR_INVALID_URL'\n  }\n}\n\nNode.js v19.1.0\n"`

**Reply by chrisgrieser** [reply]
same here. lsp installed via mason, getting the same error as kofm. And weirdly, the modeline-workaround also works for me (but breaks too much other file-type dependent stuff to be a usable workaround...)

Not sure that that error message means, disabling treesitter the markdown (and markdown_inline) parser does not seem to solve it

**Comment by DanRoscigno** [comment]
I have a minimal working config for grammarly and marksman here, maybe this will help. See the README, as you have to run  the pnpm commands or you will get the missing wasm ereor. 
 https://github.com/DanRoscigno/nvim-markdown-grammarly

**Reply by kofm** [reply]
While this may certainly work, the problem is that it works with a very minimal neovim config which does not use nvim-treesitter (which I suspect is the culprit of the issue).
If fact, if I start neovim with `nvim -c "set ft=pandoc" filename.md` it works right away without any modeline hack

**Reply by DanRoscigno** [reply]
@kofm That minimal config does not have treesitter; like you I was trying to figure out why I could not get Grammarly working with my config, so I used that minimal one to get going.  

The config that I use all day long every day does have treesitter and several other common plugins.  It is at [grammarly-nvim-dotfiles](https://github.com/DanRoscigno/grammarly-nvim-dotfiles/).  It is forked from [Aᴋsʜᴇᴛᴛ Rᴀɪ Jɪɴᴅᴀʟ](https://github.com/akshettrj/nvim-dotfiles).  I found Akshett's config by searching GitHub for `lspconfig.grammarly.setup` if I remember correctly.

**Reply by ljcolling** [reply]
I got this working ok, which is great, but I was wondering if you have any leads on how I might go about changing the documentDialect setting from 'american' to 'british'?

**Reply by ljcolling** [reply]
Well, I hacked something together and hardcoded it into the server code. Not ideal, but it gets the job done.

**Comment by DanRoscigno** [comment]
@chrisgrieser , I have seen the `grammarly-languageserver/node_modules/web-tree-sitter/tree-sitter.wasm',\n code: 'ERR_INVALID_URL`  error a few times, and each time I fixed it by running the `pnpm install`, `pnpm run build`, and `pnpm test` commands as detailed at the end of [this README](https://github.com/DanRoscigno/kickstart.nvim#adding-grammarly-and-marksman).  The README I linked to is added to my fork of TJ DeVries' recent "kickstart" config for neovim.

**Comment by Troublor** [comment]
Using the version published on Homebrew should work out of box: https://formulae.brew.sh/formula/grammarly-languageserver

**Comment by comiluv** [comment]
I solved the problem by downgrading node.js to 16.19.0

**Comment by jrslwlkn** [comment]
I got mine working with the following:

```lua
require'lspconfig'.grammarly.setup {
    filetypes = { "markdown", "tex", "text", },
    init_options = {
        clientId = "client_"
    },
    root_dir = function(fname)
        return require'lspconfig'.util.find_git_ancestor(fname) or vim.loop.os_homedir()
    end,
}
```
